lib/elixir/test/elixir/code_normalizer_test.exs

```diff
@@ -0,0 +1,364 @@
+    quoted =
+      Code.string_to_quoted!("""
+      foo(:bar) do
+        :ok
+      end
+      """)
+
+    normalized = Normalizer.normalize(quoted)
+
+    assert {:foo, _,
+            [
+              {:__block__, _, [:bar]},
+              {:__block__, _,
+               [
+                 [{{:__block__, _, [:do]}, {:__block__, _, [:ok]}}]
+               ]}
+            ]} = normalized
+  end
+
+  test "do, else, catch, rescue and after blocks" do
+    quoted =
+      Code.string_to_quoted!("""
+      foo do
+        :in_do
+      else
+        :in_else
+      catch
+        :in_catch
+      rescue
+        :in_rescue
+      after
+        :in_after
+      end
+      """)
+
+    normalized = Normalizer.normalize(quoted)
+
+    assert {:foo, _,
+            [
+              {:__block__, _,
+               [
+                 [
+                   {{:__block__, _, [:do]}, {:__block__, _, [:in_do]}},
+                   {{:__block__, _, [:else]}, {:__block__, _, [:in_else]}},
+                   {{:__block__, _, [:catch]}, {:__block__, _, [:in_catch]}},
+                   {{:__block__, _, [:rescue]}, {:__block__, _, [:in_rescue]}},
+                   {{:__block__, _, [:after]}, {:__block__, _, [:in_after]}}
+                 ]
+               ]}
+            ]} = normalized
+  end
+
+  test "left to right arrows" do
+    quoted =
+      Code.string_to_quoted!("""
+      case do
+        :left -> :right
+      end
+      """)
+
+    normalized = Normalizer.normalize(quoted)
+
+    assert {:case, _,
+            [
+              {:__block__, _,
+               [
+                 [
+                   {{:__block__, _, [:do]},
+                    [
+                      {:->, _,
+                       [
+                         [{:__block__, _, [:left]}],
+                         {:__block__, _, [:right]}
+                       ]}
+                    ]}
+                 ]
+               ]}
+            ]} = normalized
+  end
+
+  test "left to right arrow with multiple arguments" do
+    quoted =
+      Code.string_to_quoted!("""
+      case do
+        :foo, :bar -> :baz
+      end
+      """)
+
+    normalized = Normalizer.normalize(quoted)
+
+    assert {:case, _,
+            [
+              {:__block__, _,
+               [
+                 [
+                   {{:__block__, _, [:do]},
+                    [
+                      {:->, _,
+                       [
+                         [
+                           {:__block__, _, [:foo]},
+                           {:__block__, _, [:bar]}
+                         ],
+                         {:__block__, _, [:baz]}
+                       ]}
+                    ]}
+                 ]
+               ]}
+            ]} = normalized
+  end
+
+  test "unqualified call's form is skipped" do
+    quoted = Code.string_to_quoted!("foo()")
+    normalized = Normalizer.normalize(quoted)
+
+    assert {:foo, _, []} = normalized
+  end
```

Comment:
Those tests look great but they are all unit tests. Ultimately, the output of normalize is not important if the formatter can't understand it. I think more important would be to unit test that the quoted expressions get formatted properly. Perhaps by copying Macro.to_string suite (or alternatively we should change Macro.to_string suite to the new functionality, then we get all tests for free and we can remove these tests).

Reply:
Actually, we can't replace Macro.to_string because it supports an anonymous function as second argument, but we can copy the tests. :)

Reply:
I'm a bit lost on how would I use those tests. Macro.to_string can take any quoted expression, and it's output is quite different to that of the formatter. Wouldn't it be better to copy the code_formatter tests, and assert that both Code.format_string and Code.quoted_to_algebra + some_doc_to_string_function produce the same output?

I was doing something similar in the tests for my first iteration of the normalizer but I wasn't sure if it was correct: https://github.com/doorgan/formatter/blob/main/test/integration_test.exs

Reply:
Done :)